packages/cli/src/ui/hooks/vim.ts

```diff
@@ -0,0 +1,1571 @@
+                      }
+                    }
+                  }
+
+                  buffer.replaceRangeByOffset(startOffset, endOffset, '');
+                }
+                setModeImmediate('INSERT');
+              };
+              setPendingC(false);
+              setModeImmediate('INSERT');
+              return true;
+            }
+
+            // Normal up movement
+            for (let i = 0; i < repeatCount; i++) {
+              buffer.move('up');
+            }
+            clearCount();
+            return true;
+          }
+
+          case 'l': {
+            // Check if this is part of a change command (cl)
+            if (pendingC) {
+              const repeatCount = getCurrentCount();
+              clearCount();
+
+              // Change N characters to the right
+              for (let i = 0; i < repeatCount; i++) {
+                buffer.del();
+              }
+
+              // Record this command for repeat and switch to INSERT mode
+              const commandRepeatCount = repeatCount;
+              lastCommandRef.current = () => {
+                // Change N characters to the right
+                for (let i = 0; i < commandRepeatCount; i++) {
+                  const currentRow = buffer.cursor[0];
+                  const currentCol = buffer.cursor[1];
+                  const currentLine = buffer.lines[currentRow] || '';
+
+                  if (currentCol < currentLine.length) {
+                    buffer.del();
+                  }
+                }
+                setModeImmediate('INSERT');
+              };
+              setPendingC(false);
+              setModeImmediate('INSERT');
+              return true;
+            }
+
+            // Normal right movement
+            for (let i = 0; i < repeatCount; i++) {
+              const currentRow = buffer.cursor[0];
+              const currentCol = buffer.cursor[1];
+              const currentLine = buffer.lines[currentRow] || '';
+
+              // Don't move past the last character of the line
+              if (currentCol < currentLine.length - 1) {
+                buffer.move('right');
+              } else if (currentRow < buffer.lines.length - 1) {
+                // Move to beginning of next line
+                buffer.move('down');
+                buffer.move('home');
+              }
+            }
+            clearCount();
+            return true;
+          }
+
+          case 'w': {
+            // Check if this is part of a delete command (dw)
+            if (pendingD) {
+              const repeatCount = getCurrentCount();
+              clearCount();
+
+              const text = buffer.text;
+              const currentOffset = getCurrentOffset();
+              let endOffset = currentOffset;
+
+              // Delete from cursor through N words using findNextWordStart for consistency
+              let searchOffset = currentOffset;
+              for (let i = 0; i < repeatCount; i++) {
```

Comment:
this logic is not safe to perform here. instead use the existing api in text-buffer.ts or create one. that. is crucial to make sure this logic runs inside the reducer in text-buffer that ensures commands are executed properly and in order even when multiple key strokes are executed in the same react loop. the same concern applies to all of the other cases here that modify the state of the text. buffer. they need to instead call a method on text-buffer.ts to apply the transformation desired rather than reading the possibly stale state of the text buffer and performing operations on it.
sorry this part of how text-buffer works is a little surprising but it is better than the alternatives where it becomes really difficult to be confident that react is dealing with the correct state.